ipv8/REST/dht_endpoint.py

```diff
@@ -99,53 +103,65 @@ def reconstruct_all_blocks(self, block_chunks, public_key):
 
         return new_blocks, max_version
 
-    def publish_latest_block(self):
+    def publish_latest_block(self, block, attempt):
         """
         Publish the latest block of this node's TrustChain to the DHT
         """
-        latest_block = self.trustchain.persistence.get_latest(self.trustchain.my_peer.public_key.key_to_bin())
-
-        if latest_block:
-            # Get all the previously published blocks for this peer from the DHT, and check if this is a duplicate
-            latest_block = latest_block.pack()
-
-            def on_failure(failure):
-                logging.error("Publishing latest block failed with %s:\n%s", failure.value.__class__.__name__,
-                              failure.value.message)
-
-            def on_success(block_chunks):
+        if attempt > self.ATTEMPT_LIMIT:
+            logging.error("Publishing latest block failed after %d attempts", self.ATTEMPT_LIMIT)
+            return
+
+        if block:
+            latest_block = block.pack()
+
+            # Get the total number of chunks in this blocks
+            total_chunks = len(latest_block) // self.CHUNK_SIZE
+            total_chunks += 1 if len(latest_block) % self.CHUNK_SIZE != 0 else 0
+
+            def publish_chunk(_, slice_pointer, chunk_idx, chunk_attempt=1):
+                # If we've reached the end of the block, we stop the chain, and increment the block version
+                if chunk_idx >= total_chunks:
+                    self.block_version += 1
+                    return
+
+                # If we've attempted to publish this block more times than allowed, give up on the chunk and the block
+                if chunk_attempt > self.ATTEMPT_LIMIT:
+                    logging.error("Publishing latest block failed after %d attempts on chunk %d", self.ATTEMPT_LIMIT,
+                                  chunk_idx)
+                    return
+
+                # Prepare and pack the chunk for publishing to the DHT
+                my_private_key = self.trustchain.my_peer.key
+                chunk = latest_block[slice_pointer: slice_pointer + self.CHUNK_SIZE]
+                signature = my_private_key.signature(str(self.block_version).encode('utf-8')
```

Comment:
I think it's better to use the IPv8 serializer to construct the data to sign.  This would also avoid the problem of giving an attacker the ability to construct valid DHT entries by replaying signatures.

Reply:
Should be done now, I think.